packages/core/src/saml-applications/routes/utils.ts

```diff
@@ -0,0 +1,299 @@
+      user.email &&
+      user.email_verified
+    ) {
+      return {
+        NameIDFormat: format,
+        NameID: user.email,
+      };
+    }
+    // For other formats or when email is not available, use sub
+    if (format === saml.Constants.namespace.format.persistent) {
+      return {
+        NameIDFormat: format,
+        NameID: user.sub,
+      };
+    }
+  }
+  // No nameIDFormat specified, use default logic
+  // Use email if available
+  if (user.email && user.email_verified) {
+    return {
+      NameIDFormat: saml.Constants.namespace.format.emailAddress,
+      NameID: user.email,
+    };
+  }
+  // Fallback to persistent format with user.sub
+  return {
+    NameIDFormat: saml.Constants.namespace.format.persistent,
+    NameID: user.sub,
+  };
+};
+
+export const createSamlTemplateCallback =
+  (
+    idp: saml.IdentityProviderInstance,
+    sp: saml.ServiceProviderInstance,
+    user: IdTokenProfileStandardClaims
+  ) =>
+  (template: string) => {
+    const assertionConsumerServiceUrl = sp.entityMeta.getAssertionConsumerService(
+      saml.Constants.wording.binding.post
+    );
+
+    const { nameIDFormat } = idp.entitySetting;
+    const { NameIDFormat, NameID } = getSamlNameId(user, nameIDFormat);
+
+    const id = `ID_${generateStandardId()}`;
+    const now = new Date();
+    const expireAt = new Date(now.getTime() + 10 * 60 * 1000); // 10 minutes later
+
+    const tagValues = {
+      ID: id,
+      AssertionID: `ID_${generateStandardId()}`,
+      Destination: assertionConsumerServiceUrl,
+      Audience: sp.entityMeta.getEntityID(),
+      EntityID: sp.entityMeta.getEntityID(),
+      SubjectRecipient: assertionConsumerServiceUrl,
+      Issuer: idp.entityMeta.getEntityID(),
+      IssueInstant: now.toISOString(),
+      AssertionConsumerServiceURL: assertionConsumerServiceUrl,
+      StatusCode: saml.Constants.StatusCode.Success,
+      ConditionsNotBefore: now.toISOString(),
+      ConditionsNotOnOrAfter: expireAt.toISOString(),
+      SubjectConfirmationDataNotOnOrAfter: expireAt.toISOString(),
+      NameIDFormat,
+      NameID,
+      InResponseTo: 'null',
+      /**
+       * User attributes for SAML response
+       *
+       * @todo Support custom attribute mapping
+       * @see {@link https://github.com/tngan/samlify/blob/master/src/libsaml.ts#L275-L300|samlify implementation}
+       *
+       * @remarks
+       * By examining the code provided in the link above, we can define all the attributes supported by the attribute mapping here. Only the attributes defined in the `loginResponseTemplate.attributes` added when creating the IdP instance will appear in the SAML response.
+       */
+      attrSub: user.sub,
+      attrEmail: user.email,
+      attrName: user.name,
+    };
+
+    const context = saml.SamlLib.replaceTagsByValue(template, tagValues);
+
+    return {
+      id,
+      context,
+    };
+  };
+
+export const exchangeAuthorizationCode = async (
```

Comment:
nit: not sure if we could use the SSO Oidc connector methods directly here. Otherwise, we might need to revisit those dependencies from SSO OidcConnector.  It was just for a quick demo use only. For production implementation, need a better design. What is the relation between SSO Oidc connector and the SAML callback here.

Reply:
Some methods (use auth code to get access token, get OIDC open configuration) are used by both SAML app and SSO connectors, let's refactor them later. LOG-10600